### PR TITLE
Dionisia - Edges - Media Ranker

### PR DESCRIPTION
**Pull request by larachan15**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a model method for self.get_categories in the work model. This was for my create a new form so that there would be a drop down menu to choose from the three work categories: album, movie or book.
Describe how you approached testing that model method. What edge cases did you come up with? |  In the test, I tested if the categories: album, movie, and book were true based on my validations in the model. I also wanted to be sure some other category wouldn't be acceptable so I also created a test category "FAKE" which returned false.
What are `session` and `flash`? What is the difference between them? | Both are hash like objects that uses cookies. Session is used to record who is logged into a particular session. Flash is used to give a user a message to alert them if a submission was successful, or if an error occurred.
Describe a controller filter you wrote. | In my Works Controller, I made a find_work filter. This was to replace the redundancy in the show, edit, update and destroy actions.
What was one thing that you gained more clarity on through this assignment? | I think l learned more about how to retrieve and manipulate data in Active Record. I also better understand the relation between the model and the controllers and how to display information using them. I also learned how to use inspect in the browser a little better.
What is the Heroku URL of your deployed application? | https://dionisia-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | This was a pretty good learning experience, although the project was a lot. I'm not sure if there could have been a little more in class project time fit in, but that would have been nice. Or maybe the assessment test could have been handed out earlier so we could have started on it sooner.

### CODE REVIEW

**Comment by tildeee** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | x
Comprehension questions	| x, feedback accepted! Thank you
**General** |
Rails fundamentals (RESTful routing, use of named paths)  | x
Views are well-organized (DRY, use of semantic HTML, use of partials) | x
Errors are reported to the user | x
Business logic lives in the models | x, some people put the logic of getting the most-voted and the top 10 media in the `Work` model, where you put it in the `WorksController`. If you ever moved it into the model, I would expect testing around it
Models are thoroughly tested, including relations, validations and any custom logic | x, missing a few on relations?
**Wave 1 - Media** |
Splash page shows the three media categories | x
Basic CRUD operations on media are present and functional | x
**Wave 2 - Users and Votes** |
Users can log in and log out | x
The ID of the current user is stored in the session | x
A user cannot vote for the same media more than once | x
All media lists are ordered by vote count | x
Splash page contains a media spotlight | x
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | x
Individual user pages and the user list are present | x
**Optional - Styling** |
Bootstrap is used appropriately | x
Look and feel is similar to the original | x
**Overall** |

Dionisia!!!! This project looks great!! The code looks great and the project looks great! It hit all of the requirements and I'm really happy about this submission.

I really only have a few comments on this project, mostly around tests on your models around relations. Otherwise... I think the code you submitted is really clean and does some really cool things.

Great work overall!

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Work < ApplicationRecord
+  has_many :votes, dependent: :destroy
+  has_many :users, through: :votes
+  validates :title, presence: true, uniqueness: true
+  validates :category, presence: true, inclusion: { in: %w(album movie book) }
+  validates :publication_year, numericality: { only_integer: true }
+  scope :ordering, -> {order('votes_count DESC')}
```
Comment (tildeee) [inline app/models/work.rb:7]: Niiiiiice! Very interesting and cool

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  CATEGORIES = %w[album book movie]
+
+  def self.get_categories
+    return CATEGORIES
+  end
```
Comment (tildeee) [inline app/models/work.rb:13]: Nicely done!!! I really like this class method. Side note, however... Turns out that you can access constants defined in a class with the following syntax: `Work::CATEGORIES` (which would replace this class method `Work.get_categories`). You can do that in rails console to check!

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          </li>
+        </ul>
+        <% end %>
+      </nav>
+    </header>
+
+
+    <section class="flash-messages">
+      <% flash.each do |key, message| %>
+        <% alert_class = key == "error" ? "warning" : key %>
```
Comment (tildeee) [inline app/views/layouts/application.html.erb:45]: Nicely done! This little bit was tricky!

File: app/views/sessions/_session_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<%= render partial: 'layouts/form_errors',
+locals: {
+  model: @session
+}
+%>
```
Comment (tildeee) [inline app/views/sessions/_session_form.html.erb:5]: Does this get used?

File: app/views/users/_user_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<%= render partial: 'layouts/form_errors',
+locals: {
+  model: @user
+}
+%>
```
Comment (tildeee) [inline app/views/users/_user_form.html.erb:5]: Does this get used?

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+  it 'is valid when username is present' do
+    is_valid = @user.valid?
+    expect(is_valid).must_equal true
+  end
+
+  it 'is valid if username is unique' do
+    @user.user_name = User.first.user_name
+    expect(@user.valid?).must_equal false
+  end
```
Comment (tildeee) [inline test/models/user_test.rb:25]: Test on relations?

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      expect(@work.valid?).must_equal true
+
+      @work.category = "movie"
+      expect(@work.valid?).must_equal true
+
+      @work.category = "book"
+      expect(@work.valid?).must_equal true
+
+      @work.category = "FAKE"
+      expect(@work.valid?).must_equal false
```
Comment (tildeee) [inline test/models/work_test.rb:60]: NICE!
